docs/dev/relay_add_pass.rst

```diff
@@ -0,0 +1,318 @@
+.. _relay-add-pass:
+
+Adding a Compiler Pass to Relay
+===============================
+
+Compiler passes are the primary interface for both extending Relay's feature
+set and for performing optimizations on Relay programs. By writing a compiler
+pass, you can then modify the AST and/or collect information about the AST,
+depending on your goal. Indeed, some of Relay's most important "built-in"
+features (e.g., autodiff and type inference) are nothing more than compiler
+passes.
+
+At a high level, there are three key components to writing a pass:
+
+- Creating one or more C++ classes that traverse the program
+- Registering an API endpoint with the ``TVM_REGISTER_API`` macro that performs the pass
+- Wrapping the Python API hook in a neater interface
+
+To begin, we'll give an overview of the key mechanisms for writing a compiler
+pass. Then, we'll walk through a concrete example of the constant-folding
+pass in Relay.
+
+AST Traversers
+--------------
+
+The base class used to traverse Relay programs is ``ExprFunctor``. The public
+interface it provides is a ``VisitExpr`` method that takes an expression and
+zero or more arguments and returns an instance of some type. When you extend
+this class, you define the AST traversal pattern by providing implementations
+of ``VisitExpr_`` for each type of expression.
+
+``ExprFunctor`` itself is a very general class, which is why more often than
+not, you will be extending ``ExprVisitor`` or ``ExprMutator``. These classes
+extend ``ExprFunctor`` and provide default implementations of ``VisitExpr_``
+for each expression type that capture common traversal patterns. Having these
+default implementations means we only need to provide overriding
+implementations for the expression types where we want different behavior. We
+describe each subclass on its own in the following sections.
+
+Expression Visitors
+~~~~~~~~~~~~~~~~~~~
+
+``ExprVisitor`` is for passes that don't modify the program and instead
+collect information about the program. With this class, ``VisitExpr`` and the
+private counterparts return nothing. The default ``VisitExpr_``
+implementations provided by this class simply visit all of the expression's
+fields that are expressions. The default implementation for ``IfNode`` is
+shown below.
+
+.. code:: c
+
+    void ExprVisitor::VisitExpr_(const IfNode* op) {
+      this->VisitExpr(op->cond);
+      this->VisitExpr(op->true_branch);
+      this->VisitExpr(op->false_branch);
+    }
+
+Note that we're calling ``VisitExpr`` and not ``VisitExpr_`` here. This way,
+routing to the appropriate implementation is taken care of for you, and you
```

Comment:
I think we can probably mention here that calling `VisitExpr` will again dispatch to the correct or needed `VisitExpr_` depending on the node type of each component of a `IfNode. I meant we probably need to briefly talk about the functionality of `VisitExpr` as well.